### PR TITLE
Feature/gptl

### LINKED ISSUE #6: Add Support for GPTL
Add support for building with GPTL so users can insert custom timers for performance analysis purposes.

### PR DESCRIPTION
**Pull request by robgpita**
## Proposed changes

Add support for profiling with [GPTL](https://github.com/jmrosinski/GPTL).  Closes  #6 .

This pull request adds two build options for profiling with GPTL.

The first option, ENABLE_GPTL, turns on profiling with GPTL and is enabled with the ENABLE_GPTL cmake flag:
```
cmake -DENABLE_GPTL=1 ......
```
The second option, ENABLE_AUTOPROFILING, turns on GPTL's autoprofiling feature if ENABLE_GPTL is also enabled:
```
cmake -DENABLE_GPTL=1 -DENABLE_AUTOPROFILING=1 ....
```

The CI MacOS and Ubuntu build tests have been updated to install GPTL and add a build test with `-DENABLE_GPTL=1 -DENABLE_AUTOPROFILING=1`

## Types of changes

What types of changes does your code introduce?

- [ ] Bugfix (Change which fixes an issue)
- [x] New feature (Change which adds functionality)
- [ ] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [ ] Documentation Update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-c_sw/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

For now, this is a draft PR.  Once the namelist I/O is sorted out, and CI is verified to work properly, it will be ready for review.

### CODE REVIEW

**Review by LyndStringer — COMMENTED** [review]
Looks good except for the tests not passing.

File: ref/cmake/compiler_flags_Intel_Fortran.cmake [hunk 1]
```diff
@@ -15,3 +15,9 @@ set( CMAKE_Fortran_FLAGS_RELEASE "-fno-alias -auto -align array64byte -xHOST -qn
 ####################################################################
 
 set( CMAKE_Fortran_FLAGS_DEBUG   "-g -O0 -warn -debug -nolib-inline -fno-inline-functions -assume protect_parens,minus0 -prec-div -prec-sqrt -check bounds -check uninit -fp-stack-check -init=snan,array " )
+
+####################################################################
+# FLAGS FOR AUTOPROFILING
+####################################################################
+
+set( Fortran_AUTOPROFILING_FLAGS        "-finstrument-functions" )
```
Comment (christopherwharrop-noaa) [inline ref/cmake/compiler_flags_Intel_Fortran.cmake]: Need a newline here.

File: ref/src/yppm_driver.F90 [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  
+    ! Turn off GPTL if enabled
+#ifdef ENABLE_GPTL
+  if (do_profile == 1) then
+    ret = gptlpr(0)
+    ret = gptlfinalize()
+  end if
+#endif
+  
+end program yppm_driver
```
Comment (christopherwharrop-noaa) [inline ref/src/yppm_driver.F90]: Need a newline here

File: ref/cmake/FindGPTL.cmake [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    endif()
+    if(GPTL_LINK_LIBRARIES)
+        set_property(TARGET GPTL::GPTL PROPERTY INTERFACE_LINK_LIBRARIES ${GPTL_LINK_LIBRARIES})
+    else()
+        set_property(TARGET GPTL::GPTL PROPERTY INTERFACE_LINK_LIBRARIES ${GPTL_LIBRARY})
+        get_filename_component(_lib_dir ${GPTL_LIBRARY} DIRECTORY)
+        set_property(TARGET GPTL::GPTL APPEND PROPERTY INTERFACE_LINK_DIRECTORIES ${_lib_dir})
+        unset(_lib_dir)
+    endif()
+endif()
```
Comment (christopherwharrop-noaa) [inline ref/cmake/FindGPTL.cmake]: Need a newline here

**Review by christopherwharrop-noaa — CHANGES_REQUESTED** [review]
A few style newline nitpicks.  Also, I believe you need to add the link libraries in the `test/CMakeLists.txt` file because the interpolation test will need them.

```
if ( GPTL_FOUND )
   target_link_libraries( ${PROJECT_NAME} PUBLIC GPTL::GPTL )
endif()
```

I think that is why the tests are failing

File: ref/test/CMakeLists.txt [hunk 4]
```diff
@@ -63,6 +63,9 @@ ${PROJECT_SOURCE_DIR}/test/src/interpolateTest.f90
 )
 add_executable (interpolateTest ${test_src_files}  ${PROJECT_SOURCE_DIR}/src/interpolate.f90)
 
+if ( GPTL_FOUND )
+   target_link_libraries( ${PROJECT_NAME} PUBLIC GPTL::GPTL )
```
Comment (christopherwharrop-noaa) [inline ref/test/CMakeLists.txt]: I think you need to replace `${PROJECT_NAME}` with `interpolateTest` here.  Sorry I wasn't clear that the name of the target needed to be the name of the executable in question, which in this case is `interpolateTest`.

**Review by christopherwharrop-noaa — APPROVED** [review]
Thank you for tackling this.

**Comment by christopherwharrop-noaa** [comment]
@LyndStringer & @middlecoff - The GPTL for yppm is no longer a Draft PR and is now ready for a proper review.

**Review by LyndStringer — APPROVED** [review]
Looks good.

**Pull request merged**